### PR TITLE
feat(adapters): knowledge-bundle context sources (#736, #763, #767, #776)

### LINKED ISSUE #736: Add OKF bundle loader as a context source
## Context

OKF is a good fit for `contextweaver` as a portable knowledge source: OKF stores concepts as Markdown files with YAML frontmatter; `contextweaver` should decide what subset enters the prompt through its existing candidate selection, budget, dedup, sensitivity, and rendering pipeline.

## Goal

Allow `contextweaver` to ingest an OKF directory and expose OKF concepts as selectable context candidates.

## Scope

- Add an OKF loader module, for example `src/contextweaver/adapters/okf.py` or `src/contextweaver/sources/okf.py` depending on the current package layout.
- Parse `.md` files with YAML frontmatter.
- Treat `index.md` and `log.md` according to OKF conventions:
  - `index.md` as optional overview / bundle metadata.
  - `log.md` as optional bundle history, not normal concept content by default.
- Convert concept files into internal context candidates / `ContextItem`-compatible objects.
- Use known fields when present: `type`, `title`, `description`, `resource`, `tags`, `timestamp`, `links`.
- Preserve unknown frontmatter fields as metadata.
- Add deterministic ordering and stable IDs.
- Add fixtures under `tests/fixtures/okf/`.
- Add a recipe doc: `docs/recipes/okf_bundle.md`.

## Design constraints

- Do not require OKF in the core runtime path.
- Keep parsing permissive: unknown fields and unknown types should not fail ingestion.
- Treat invalid frontmatter as a recoverable parse error where possible, with diagnostics.
- Avoid overfitting to Google Cloud-specific examples; implement the generic OKF shape.

## Acceptance criteria

- [ ] A small OKF bundle can be loaded into `contextweaver` without network access.
- [ ] Loaded concepts can be ranked/selected/rendered by the existing context pipeline.
- [ ] Unknown OKF fields are preserved in metadata.
- [ ] `index.md` and `log.md` have documented behavior.
- [ ] Tests cover: valid bundle, unknown type, missing title, broken link, invalid frontmatter.
- [ ] Docs include a minimal example and explain when to prefer OKF vs normal event-log ingestion.

## Labels / milestone

Suggested labels: `enhancement`, `interop`.
Suggested milestone: `OKF interoperability` if/when a milestone is created.

### PR DESCRIPTION
**Pull request by dgenio**
## Summary

Adds four OKF-style Markdown-plus-YAML-frontmatter knowledge-source adapters, built on one shared permissive parsing core, so external knowledge flows through contextweaver's existing candidate-selection/budget/dedup/sensitivity/rendering pipeline unchanged. No pipeline stage was modified; no new `ItemKind` was introduced — nodes materialise as `doc_snippet` with provenance under the `_contextweaver` metadata namespace.

Fixes #736, #763, #767, #776

This group was selected via a full triage pass over the repo's 175 open issues (2 closed as already-fixed in that pass: #743, #745): the four issues here share an explicit, author-declared dependency chain (#763/#767/#776 each cite #736 as prerequisite work) and one implementation path — a generic OKF loader plus three typed consumers — making them the most coherent single-PR candidate in the backlog, ahead of a runner-up CI-hygiene batch (#747/#748/#754/#757).

## Changes

- **`adapters/okf.py` (#736)** — `load_okf_bundle` parses an OKF concept directory; `index.md`/`log.md` are recognised bundle metadata/history, excluded from concept content by default. `select_knowledge` ranks and packs nodes deterministically against a query and token budget.
- **`adapters/repo_knowledge.py` (#763)** — narrows the OKF loader to generated repo docs: plain Markdown without frontmatter still loads (filename-derived title), `max_files`/`max_total_bytes` guardrails bound large trees, and `classify_usage` assigns deterministic usage tags (`"debugging"`, `"onboarding"`, ...) — plain metadata strings, **not** `Phase` enum values (extending `Phase` is issue #587's separate concern). Links are never auto-followed.
- **`adapters/lessons.py` (#767)** — adds lifecycle-aware eligibility on top of the shared core: `LessonSelectionPolicy` excludes `rejected`/`deprecated` lessons and unreviewed `candidate` lessons by default (opt in via `include_candidates=True`); every exclusion is reported with an explicit reason, never silently dropped.
- **`adapters/expertise_pack.py` (#776)** — validates pack *structure* (an `index.md` declaring `version`, every node carrying a `key`) rather than the full external weaver-spec schema (`dgenio/weaver-spec#184`, tracked as a follow-up seam — out of this session's repo scope to fetch). Adds deterministic (non-LLM) conflict detection: constraints sharing a `key` that disagree on text, scoped by `task_tags` and liveness.
- **`adapters/_okf_io.py` + `_okf_materialize.py`** — private shared core (permissive frontmatter parsing that never raises, `KnowledgeNode` dataclass, deterministic directory walk, materialisation into `ContextItem` + relevance scoring reusing `_utils.tokenize`/`jaccard`). **Not** re-exported at the `contextweaver.adapters` package level — `adapters/__init__.py` is a frozen, grandfathered module already at its size ceiling, so this mirrors how the `mcp_primitives`/`gateway_primitives` family is deliberately kept at submodule level. Import directly, e.g. `from contextweaver.adapters.okf import load_okf_bundle`.
- **Tests** — 58 new tests across `tests/test_adapters_{okf,repo_knowledge,lessons,expertise_pack}.py` covering every acceptance-criteria edge case from all four issues (valid bundle / unknown type / missing title / broken link / invalid frontmatter; all six lesson lifecycle states; missing / malformed / stale / conflicting packs), plus determinism and `to_dict()` round-trip checks. Fixtures under `tests/fixtures/{okf,lessons,expertise}/`.
- **Docs/examples** — `docs/recipes/okf_bundle.md` (indexed in `mkdocs.yml` nav and `llms.txt`/`llms-full.txt`), `examples/knowledge_bundles_demo.py` wired into `make example`, `AGENTS.md` module map, `CHANGELOG.md`.

## Checklist

- [x] Tests added or updated for every new/changed public function
- [x] `make ci` passes locally (fmt + lint + type + test + drift-check + module-size-check + doc-snippets-check + readme-version-check + example + demo) — verified target-by-target since `uv`/full toolchain wasn't preinstalled in this session; only a pre-existing, network-sandboxed tiktoken test failure remains, reproduced identically against the unmodified tree
- [x] `CHANGELOG.md` updated under `## [Unreleased]`
- [x] Docstrings added for all new public APIs (Google-style)
- [x] Public-API change? — no `contextweaver.adapters.__all__` change, so `api/public_api.txt` has no drift (verified); see note above on why the new modules stay at submodule level
- [x] Every modified module stays ≤ 300 lines (`_okf_io.py` 299, `_okf_materialize.py` 100, `okf.py` 176, `repo_knowledge.py` 204, `lessons.py` 214, `expertise_pack.py` 260 — `make module-size-check` passes with no baseline changes)
- [x] Related issue linked in the summary above
- [x] Agent-facing docs updated (`AGENTS.md` module map)

## Notes for reviewers

- **#776's conflict detection is deliberately scoped down** from the issue's literal wording ("Highlight conflicts between the pack and current user/task context") to deterministic, literal same-`key` text disagreement — no natural-language contradiction inference, since that would require an LLM call the core knowledge-source loading path doesn't make. Documented in the module docstring as a follow-up seam.
- **#763's "classify by usage phase"** is implemented as plain metadata tags, not by extending the `Phase` enum, to avoid colliding with the separate open issue #587 (which is specifically about extending `Phase`).
- **Out of scope, left as natural follow-ups**: #737 (the reverse OKF-export direction), #764 (validation/drift checks), #765 (visualizer prototype), #766 (offline bundle generator) — all depend on this loader landing first.
- `.venv-dev` was used locally for verification (this sandbox had no `uv`/toolchain preinstalled) and is not part of this diff.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
_Generated by [Claude Code](https://claude.ai/code/session_01X21ymBbGbr9UcxjZQgSUfJ)_

### CODE REVIEW

File: src/contextweaver/adapters/okf.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+        msg = f"load_okf_bundle: not a directory: {root}"
+        raise ConfigError(msg)
+
+    diagnostics: list[LoadDiagnostic] = []
+    nodes: list[KnowledgeNode] = []
+    for file_path in discover_concept_files(root):
+        source_path = file_path.relative_to(root).as_posix()
+        node, diagnostic = node_from_markdown(
+            file_path.read_text(encoding="utf-8"), source_path=source_path
+        )
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/contextweaver/adapters/okf.py:95]: Good catch. Fixed in f90a4e5 — added a shared `read_markdown_text()` helper in `_okf_io.py` (decodes via `errors="replace"`) and switched `load_okf_bundle` to use it.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/contextweaver/adapters/okf.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+
+def _load_bundle_file(root: Path, filename: str) -> KnowledgeNode | None:
+    """Load a bundle-level file (``index.md`` / ``log.md``) if present."""
+    file_path = root / filename
+    if not file_path.is_file():
+        return None
+    node, _diagnostic = node_from_markdown(
+        file_path.read_text(encoding="utf-8"), source_path=filename
+    )
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/contextweaver/adapters/okf.py]: Fixed in f90a4e5 — `_load_bundle_file` now uses the same shared `read_markdown_text()` helper.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/contextweaver/adapters/lessons.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+        msg = f"load_lesson_bundle: not a directory: {root}"
+        raise ConfigError(msg)
+
+    diagnostics: list[LoadDiagnostic] = []
+    nodes: list[KnowledgeNode] = []
+    for file_path in discover_concept_files(root):
+        source_path = file_path.relative_to(root).as_posix()
+        node, diagnostic = node_from_markdown(
+            file_path.read_text(encoding="utf-8"), source_path=source_path
+        )
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/contextweaver/adapters/lessons.py:102]: Fixed in f90a4e5 — `load_lesson_bundle` now uses the shared `read_markdown_text()` helper from `_okf_io.py`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/contextweaver/adapters/expertise_pack.py [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+
+    diagnostics: list[LoadDiagnostic] = []
+    version = _load_pack_version(root, diagnostics, on_invalid=on_invalid)
+
+    nodes: list[KnowledgeNode] = []
+    for file_path in discover_concept_files(root):
+        source_path = file_path.relative_to(root).as_posix()
+        node, diagnostic = node_from_markdown(
+            file_path.read_text(encoding="utf-8"), source_path=source_path
+        )
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/contextweaver/adapters/expertise_pack.py:129]: Fixed in f90a4e5 — `load_expertise_pack` now uses the shared `read_markdown_text()` helper.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/contextweaver/adapters/expertise_pack.py [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+        _record(
+            diagnostics,
+            LoadDiagnostic(level="warning", path=VERSION_FILENAME, message="missing index.md"),
+            on_invalid=on_invalid,
+            label="load_expertise_pack",
+        )
+        return None
+    node, diagnostic = node_from_markdown(
+        index_path.read_text(encoding="utf-8"), source_path=VERSION_FILENAME
+    )
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/contextweaver/adapters/expertise_pack.py:171]: Fixed in f90a4e5 — `_load_pack_version` now uses the shared `read_markdown_text()` helper too, so version parsing is aligned with the rest of the family.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/contextweaver/adapters/repo_knowledge.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+def classify_usage(node: KnowledgeNode) -> list[str]:
+    """Return deterministic usage tags for *node* (e.g. ``["debugging"]``).
+
+    Matches keywords against the node's type, title, and tags. A node may
+    match zero, one, or several tags; matches are returned in
+    :data:`_USAGE_KEYWORDS` order, never randomised.
+    """
+    haystack = " ".join([node.node_type, node.title, *node.tags]).lower()
+    return [tag for tag, keywords in _USAGE_KEYWORDS if any(kw in haystack for kw in keywords)]
+
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/contextweaver/adapters/repo_knowledge.py:92]: Fixed in f90a4e5 — `classify_usage` now matches on a word boundary (`\b` + keyword) instead of raw substring, so "test" no longer matches inside "latest" while "debug" still matches "debugging" as intended. Added tests for both cases.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/contextweaver/adapters/repo_knowledge.py [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+    packed: list[ContextItem] = []
+    remaining = budget_tokens
+    for node in ranked:
+        item = node_to_context_item(node, source_kind=SOURCE_KIND, estimator=estimator)
+        cost = max(1, int(item.token_estimate))
+        if cost > remaining:
+            continue
+        packed.append(item)
+        remaining -= cost
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/contextweaver/adapters/repo_knowledge.py:213]: Fixed in f90a4e5 — factored the usage-tag stamping into a shared `_to_context_item_with_usage_tags()` helper and switched `select_repo_knowledge` to use it, so both materialisation paths now produce the same metadata shape.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/contextweaver/adapters/expertise_pack.py [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+
+    Expired nodes and (when *task_tags* is given) inapplicable nodes are
+    excluded — pack sections enter context "only when relevant" (issue #776
+    acceptance criteria), never unconditionally.
+    """
+    return [
+        node_to_context_item(node, source_kind=SOURCE_KIND, estimator=estimator)
+        for node in pack.nodes
+        if not node.is_expired(now=now) and _applies(node, task_tags)
+    ]
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/contextweaver/adapters/expertise_pack.py:258]: Fixed in f90a4e5 — `expertise_pack_to_context_items` now excludes nodes missing `key` (the ones already flagged at load time), so a structurally-invalid node can no longer enter bounded context.

---
_Generated by [Claude Code](https://claude.ai/code)_
